vm-virtio/src/queue.rs

```diff
@@ -29,6 +29,37 @@ pub(super) const VIRTQ_DESC_F_WRITE: u16 = 0x2;
 // The Virtio Spec 1.0 defines the alignment of VirtIO descriptor is 16 bytes,
 // which fulfills the explicit constraint of GuestMemoryMmap::read_obj().
 
+/// An iterator over a single descriptor chain.  Not to be confused with AvailIter,
+/// which iterates over the descriptor chain heads in a queue.
+pub struct DescIter<'a> {
+    next: Option<DescriptorChain<'a>>,
+}
+
+impl<'a> DescIter<'a> {
+    /// Returns an iterator that only yields the readable descriptors in the chain.
+    pub fn readable(self) -> impl Iterator<Item = DescriptorChain<'a>> {
+        self.skip_while(DescriptorChain::is_write_only)
+    }
```

Comment:
This is also incorrect because it will include the writable descriptors as well.  `skip_while()` returns all elements after the closure returns false and the closure will always return false on the first element because the readable descriptors have to appear in the descriptor chain before any writable ones.

Again, I would suggest changing this to either:

```rust
self.take_while(DescriptorChain::is_read_only)
```

(which is how this function is implemented in crosvm).  Or 

```rust
self.filter(|d| !d.is_write_only())
```

or

```rust
self.take_while(|d| !d.is_write_only())
```

Reply:
:+1: